src/python/foglamp/sending_process.py

```diff
@@ -0,0 +1,813 @@
+        raise
+
+    return data_to_send
+
+
+def last_object_id_read():
+    """ Retrieves the starting point for the send operation
+
+    Returns:
+        last_object_id: starting point for the send operation
+
+    Raises:
+    Todo:
+        it should evolve using the DB layer
+    """
+
+    global _pg_cur
+
+    try:
+        sql_cmd = "SELECT last_object FROM foglamp.streams WHERE id={0}".format(_stream_id)
+
+        _pg_cur.execute(sql_cmd)
+        rows = _pg_cur.fetchall()
+
+        if len(rows) == 0:
+            _message = _MESSAGES_LIST["e000016"].format(str(_stream_id))
+            raise ValueError(_message)
+
+        elif len(rows) > 1:
+
+            _message = _MESSAGES_LIST["e000014"].format(str(_stream_id))
+            raise ValueError(_message)
+
+        else:
+            last_object_id = rows[0][0]
+            _logger.debug("DB row last_object_id |{0}| : ".format(last_object_id))
+
+    except Exception:
+        _message = _MESSAGES_LIST["e000019"]
+
+        _logger.error(_message)
+        raise
+
+    return last_object_id
+
+
+def is_stream_id_valid(stream_id):
+    """ Checks if the provided stream id  is valid
+
+    Returns:
+        True/False
+    Raises:
+    Todo:
+        it should evolve using the DB layer
+    """
+
+    global _pg_cur
+
+    try:
+        sql_cmd = "SELECT id FROM foglamp.streams WHERE id={0}".format(stream_id)
+
+        _pg_cur.execute(sql_cmd)
+        rows = _pg_cur.fetchall()
+
+        if len(rows) == 0:
+            _message = _MESSAGES_LIST["e000016"].format(str(stream_id))
+            raise ValueError(_message)
+
+        elif len(rows) > 1:
+
+            _message = _MESSAGES_LIST["e000014"].format(str(stream_id))
+            raise ValueError(_message)
+        else:
+            stream_id_valid = True
+
+    except Exception as e:
+        _message = _MESSAGES_LIST["e000013"].format(str(e))
+
+        _logger.error(_message)
+        raise e
+
+    return stream_id_valid
+
+
+def last_object_id_update(new_last_object_id):
+    """ Updates reached position
+
+    Args:
+        new_last_object_id: Last row id already sent
+
+    Todo:
+        it should evolve using the DB layer
+
+    """
+    global _pg_cur
+    global _pg_conn
+
+    try:
+        _logger.debug("Last position, sent |{0}| ".format(str(new_last_object_id)))
+
+        sql_cmd = "UPDATE foglamp.streams SET last_object={0}, ts=now()  WHERE id={1}" \
+            .format(new_last_object_id, _stream_id)
+
+        _pg_cur.execute(sql_cmd)
+
+        _pg_conn.commit()
+
+    except Exception:
+        _message = _MESSAGES_LIST["e000020"]
+
+        _logger.error(_message)
+        raise
+
+
+@_performance_log
+def _send_data_block():
+    """ Sends a block of data to the destination using the configured plugin
+
+    Args:
+    Returns:
+    Raises:
+    Todo:
+    """
+
+    data_sent = False
+    try:
+        _logger.debug("_send_data_block")
+
+        last_object_id = last_object_id_read()
+
```

Comment:
As per the specs "**A single send process is created for each destination to which readings may be sent.**". That means a single send process can send data to same destination but on multiple streams. This code works per stream, that means a stream_id is set globally and is used throughout the code. Also If we look into the method **_load_data_into_memory** it has different handlers for different type of data source but using the same last_object_id (fetched from a single stream). Is that correct and as per the specs @MarkRiddoch , @izoratti ?

Reply:
Spe was wrong here, it should have send per stream not per destination